pkg/util/exec/fake_exec.go

```diff
@@ -143,3 +145,69 @@ func (fake *FakeExitError) Exited() bool {
 func (fake *FakeExitError) ExitStatus() int {
 	return fake.Status
 }
+
+func commandAsStringSlice(command interface{}) []string {
+	if args, ok := command.([]string); ok {
+		return args
+	} else if commandstr, ok := command.(string); ok {
+		return strings.Split(commandstr, " ")
+	}
+	panic("command must be []string or string")
+}
+
+func outputAsBytes(output interface{}) []byte {
+	if outputBytes, ok := output.([]byte); ok {
+		return outputBytes
+	} else if outputStr, ok := output.(string); ok {
+		return []byte(outputStr)
+	} else if output == nil {
+		return nil
+	}
+	panic("output must be []byte or string")
+}
+
+func (fake *FakeExec) expectFakeCmd(command interface{}, fcmd *FakeCmd) {
+	expectedArgv := commandAsStringSlice(command)
+	fake.CommandScript = append(fake.CommandScript,
+		func(cmd string, args ...string) Cmd {
+			InitFakeCmd(fcmd, cmd, args...)
+			if !reflect.DeepEqual(expectedArgv, fcmd.Argv) {
+				panic(fmt.Sprintf("Wrong Exec: expected %v got %v", expectedArgv, fcmd.Argv))
+			}
+			return fcmd
+		},
+	)
+}
+
+// ExpectCombinedOutput "predicts" a call to fake.Command(...).CombinedOutput() with the
+// given command, and provides the result of that call.
+// command is either a string (command line) or a []string (argv) indicating the expected command.
+// output and err will be returned as the result of CombinedOutput() on the command.
+// output can be either a []byte or a string (which will be converted to a []byte).
+func (fake *FakeExec) ExpectCombinedOutput(command interface{}, output interface{}, err error) {
+	fake.expectFakeCmd(command, &FakeCmd{
+		CombinedOutputScript: []FakeCombinedOutputAction{
+			func() ([]byte, error) { return outputAsBytes(output), err },
+		},
+	})
+}
+
+// ExpectRun "predicts" a call to fake.Command(...).Run() with the
+// given command, and provides the result of that call.
+// command is either a string (command line) or a []string (argv) indicating the expected command.
+// stdout, stderr, and err will be returned as the result of Run() on the command.
+// stdout and stderr can be either a []byte or a string (which will be converted to a []byte).
+func (fake *FakeExec) ExpectRun(command interface{}, stdout interface{}, stderr interface{}, err error) {
+	fake.expectFakeCmd(command, &FakeCmd{
+		RunScript: []FakeRunAction{
+			func() ([]byte, []byte, error) { return outputAsBytes(stdout), outputAsBytes(stderr), err },
+		},
+	})
+}
+
+// AssertExpectedCommands ensures that all of the commands added to fake via ExpectCombinedOutput and ExpectRun were actually executed
+func (fake *FakeExec) AssertExpectedCommands() {
```

Comment:
FakeExec already panics in one case

    if fake.CommandCalls > len(fake.CommandScript)-1 {
    	panic(fmt.Sprintf("ran out of Command() actions. Could not handle command [%d]: %s args: %v", fake.CommandCalls, cmd, args))
    }

So I just copied that behavior for the new failures.

Returning a bool and letting the caller test makes the API more difficult to use again...

Originally, I wrote it with FakeExec have a `*testing.T` field which the caller sets when constructing it, and then (assuming that field is set) it calls t.Fatalf instead of panic on error (in all of the cases). The problem with that is that when a test fails, the error shows it failing at `fake_exec.go:175` or whatever, rather than `iptables_test.go:209`. panic is actually especially nice for exec errors because the full stack trace lets you see the relevant line numbers of both the test program and the code being tested. (Though I guess with a few calls to some functions in `runtime` we could provide a more detailed t.Fatalf message...)

Does passing a `*testing.T` to fakeexec sound ok?